File: .idea/app/src/main/java/com/chevstrap/rbx/UI/Elements/Settings/Pages/IntegrationsFragment.java
package com.chevstrap.rbx.UI.Elements.Settings.Pages;

import android.graphics.Color;
import android.graphics.drawable.GradientDrawable;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.chevstrap.rbx.App;
import com.chevstrap.rbx.ConfigManager;
import com.chevstrap.rbx.Extensions.CustomUIComponents;
import com.chevstrap.rbx.Models.MethodPair;
import com.chevstrap.rbx.R;
import com.chevstrap.rbx.UI.Elements.CustomDialogs.ServerHistoryFragment;

import org.json.JSONArray;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.atomic.AtomicBoolean;

public class IntegrationsFragment extends Fragment {
	ConfigManager manager;
	private LinearLayout linear1;
	private final Map<String, CustomUIComponents.ToggleResult> toggleMap = new HashMap<>();
    private final Map<String, List<CustomUIComponents.ToggleResult>> childMap = new HashMap<>();
	private TextView textview_description;

	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
		View view = inflater.inflate(R.layout.a_normal_page_fragment, container, false);
		manager = ConfigManager.getInstance();
		textview_description = view.findViewById(R.id.textview_description);

		boolean isDark = Objects.equals(App.getChevstrapSettings().getAppThemeInApp(), "dark");
		if (isDark) {
			textview_description.setTextColor(Color.parseColor("#D6D6D6"));
		} else {
			textview_description.setTextColor(Color.parseColor("#282828"));
		}

		try {
			initialize(savedInstanceState, view);
			initializeLogic();
		} catch (Exception ignored) {}

		App.setSavedFragmentActivity(this.requireActivity());
		return view;
	}
    @Override
    public void onDestroyView() {
        super.onDestroyView();
        for (CustomUIComponents.ToggleResult toggle : toggleMap.values()) {
            if (toggle.toggleSwitch != null) {
                toggle.toggleSwitch.setOnClickListener(null);
            }
        }
        toggleMap.clear();
        childMap.clear();
    }

    private void initialize(Bundle savedInstanceState, View view) {
		linear1 = view.findViewById(R.id.linear1);
	}

	private void initializeLogic() {
		AstyleButton1(linear1);
	}

	public void addSection(ViewGroup parent, String text) {
		CustomUIComponents.SectionResult section = CustomUIComponents.addSection(getContext(), text);
		parent.addView(section.sectionContainer);
	}

	public void addDivider(ViewGroup parent) {
		CustomUIComponents.DividerResult divider = CustomUIComponents.addDivider(getContext());
		parent.addView(divider.dividerView);
	}

	public void addImportButton(String name, String description, LinearLayout parent, String typeCommand) {
		CustomUIComponents.ButtonResult getCustomButton = CustomUIComponents.addImportButton(getContext(), name, description, parent);
		parent.addView(getCustomButton.buttonView);

		getCustomButton.buttonOne.setOnClickListener(v -> {
			if ("open_server_history".equals(typeCommand)) {
				ServerHistoryFragment fragment = new ServerHistoryFragment();
				fragment.show(requireActivity().getSupportFragmentManager(), "ServerHistoryDialog");
			}
		});
	}

	public void addToggle(String id, String name, String description, LinearLayout parent, MethodPair method, String controlledParentId) {
		CustomUIComponents.ToggleResult toggle = CustomUIComponents.addToggle(getContext(), name, description, parent, method);
		AtomicBoolean isCheckedG = new AtomicBoolean(false);

		if (controlledParentId != null) {
			if (!childMap.containsKey(controlledParentId)) {
				childMap.put(controlledParentId, new ArrayList<>());

				CustomUIComponents.ToggleResult parentToggle = toggleMap.get(controlledParentId);
				if (parentToggle != null) {
					boolean parentChecked = (Boolean) parentToggle.toggleSwitch.getTag();
					if (parentChecked) {
						if (toggle.enable != null) toggle.enable.run();
					} else {
						if (toggle.disable != null) toggle.disable.run();
					}
					isCheckedG = new AtomicBoolean(parentChecked);
				}
			}
			Objects.requireNonNull(childMap.get(controlledParentId)).add(toggle);
		} else {
			toggleMap.put(id, toggle);
		}

		AtomicBoolean finalIsCheckedG1 = isCheckedG;
		Runnable updateChildren = () -> {
			List<CustomUIComponents.ToggleResult> children = childMap.get(id);
			if (children != null) {
				for (CustomUIComponents.ToggleResult child : children) {
					if (finalIsCheckedG1.get()) {
						if (child.enable != null) child.enable.run();
					} else {
						if (child.disable != null) child.disable.run();
					}
				}
			}
		};

		AtomicBoolean finalIsCheckedG = isCheckedG;
		toggle.onToggleClick = (view, isChecked) -> {
			finalIsCheckedG.set(isChecked);
			updateChildren.run();
		};

		updateChildren.run();
		parent.addView(toggle.toggleView);
	}

	public void addButton(String name, String description, LinearLayout parent, String command, String typeCommand) {
		CustomUIComponents.ButtonResult button = CustomUIComponents.addButton(getContext(), name, description, parent, command, typeCommand);
		parent.addView(button.buttonView);
	}

	public void addTextbox(String name, String description, LinearLayout parent, MethodPair method, String typeIs) {
		CustomUIComponents.TextboxResult textbox = CustomUIComponents.addTextbox(getContext(), name, description, parent, method, typeIs);
		parent.addView(textbox.textboxView);
	}

	public void addDropdown(String name, String description, LinearLayout parent, final JSONArray jsonArray, MethodPair method) {
		CustomUIComponents.DropdownResult dropdown = CustomUIComponents.addDropdown(getContext(), name, description, parent, jsonArray, method);
		parent.addView(dropdown.dropDownView);
	}

	@Override
	public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
		LinearLayout parentLayout = view.findViewById(R.id.linear2);
		textview_description.setText(App.getTextLocale(App.getAppContext(), R.string.menu_integrations_description));
		Integrations.addEveryPresets(requireContext(), parentLayout, this);
	}

	private void AstyleButton1(LinearLayout button) {
		GradientDrawable drawable = new GradientDrawable();
		drawable.setCornerRadius(30);
		drawable.setColor(Color.parseColor("#151515"));
		button.setBackground(drawable);
	}
}

File: app/src/main/java/com/chevstrap/rbx/ui/views/settings/pages/IntegrationsFragment.kt
package com.chevstrap.rbx.ui.views.settings.pages

import android.graphics.drawable.GradientDrawable
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import androidx.core.graphics.toColorInt
import androidx.fragment.app.Fragment
import com.chevstrap.rbx.App
import com.chevstrap.rbx.ConfigManager
import com.chevstrap.rbx.R
import com.chevstrap.rbx.extensions.ResourceManagerEx
import com.chevstrap.rbx.ui.components.ComponentUtils
import com.chevstrap.rbx.ui.components.holders.DividerResult
import com.chevstrap.rbx.ui.components.holders.DropdownResult
import com.chevstrap.rbx.ui.components.holders.SectionResult
import com.chevstrap.rbx.ui.components.holders.TextboxResult
import com.chevstrap.rbx.ui.components.holders.ToggleResult
import com.chevstrap.rbx.ui.viewModels.MethodAction
import com.chevstrap.rbx.ui.viewModels.MethodPair
import org.json.JSONArray

class IntegrationsFragment : Fragment() {
    var manager: ConfigManager? = null
    private var linear1: LinearLayout? = null
    private val toggleMap: MutableMap<String, ToggleResult> = mutableMapOf()
    private val childMap: MutableMap<String, MutableList<ToggleResult>> = mutableMapOf()
    private val reverseToggleMap = mutableMapOf<ToggleResult, String>()
    private val dropdownMap: MutableMap<String, DropdownResult> = mutableMapOf()
    private val childDropdownMap: MutableMap<String, MutableList<DropdownResult>> = mutableMapOf()
    private var textviewDesc: TextView? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val view: View = inflater.inflate(R.layout.page_fragment, container, false)
        manager = ConfigManager.instance
        try {
            initialize(view)
            initializeLogic()
        } catch (_: Exception) {
        }

        App.savedFragmentActivity = this.requireActivity()
        return view
    }

    override fun onDestroyView() {
        super.onDestroyView()

        for (toggle in toggleMap.values) {
            toggle.toggleSwitch?.setOnClickListener(null)
        }

        toggleMap.clear()
        childMap.clear()
        reverseToggleMap.clear()

        dropdownMap.clear()
        childDropdownMap.clear()
    }

    private fun initialize(view: View) {
        linear1 = view.findViewById(R.id.linear1)
        textviewDesc = view.findViewById(R.id.textview_desc)
    }

    private fun initializeLogic() {
        textviewDesc?.text = ResourceManagerEx.getStringOrEmpty(requireContext(), R.string.menu_integrations_description)
        linear1?.let { aStyleButton1(it) }
    }

    fun addSection(parent: ViewGroup, text: String?) {
        val section: SectionResult =  ComponentUtils.addSection(context, text)
        parent.addView(section.sectionContainer)
    }

    fun addDivider(parent: ViewGroup) {
        val divider: DividerResult =  ComponentUtils.addDivider(context)
        parent.addView(divider.dividerView)
    }

    fun addButton(
        name: String?,
        description: String?,
        parent: LinearLayout,
        method: MethodAction,
    ) {

        val button = ComponentUtils.addButton(
            context,
            name,
            description,
            method
        )

        parent.addView(button.buttonView)
    }

    private fun propagateChildren(toggleId: String?) {
        if (toggleId == null) return

        val parentToggle = toggleMap[toggleId] ?: return
        val enabled = parentToggle.isChecked

        childMap[toggleId]?.forEach { child ->
            if (enabled) {
                child.enable?.run()
            } else {
                child.disable?.run()
            }

            val childId = reverseToggleMap[child]
            propagateChildren(childId)
        }

        childDropdownMap[toggleId]?.forEach { childDropdown ->
            if (enabled) {
                childDropdown.enable?.run()
            } else {
                childDropdown.disable?.run()
            }
        }
    }

    fun addToggle(
        id: String?,
        name: String?,
        description: String?,
        parent: LinearLayout,
        method: MethodPair<*, Boolean>?,
        controlledParentId: String?
    ) {
        method ?: return

        val toggle =  ComponentUtils.addToggle(
            context,
            name,
            description,
            method
        )

        if (id != null) {
            toggleMap[id] = toggle
            reverseToggleMap[toggle] = id
        }

        if (controlledParentId != null) {
            childMap.getOrPut(controlledParentId) {
                mutableListOf()
            }.add(toggle)
        }

        val originalEnable = toggle.enable
        val originalDisable = toggle.disable

        toggle.enable = Runnable {
            originalEnable?.run()
            propagateChildren(id)
        }

        toggle.disable = Runnable {
            originalDisable?.run()
            propagateChildren(id)
        }

        toggle.onToggleClick = ToggleResult.OnToggleClickListener { _, checked ->
            toggle.isChecked = checked
            propagateChildren(id)
        }

        if (controlledParentId != null) {
            val parentToggle = toggleMap[controlledParentId]

            if (parentToggle?.isChecked == true) {
                originalEnable?.run()
            } else {
                originalDisable?.run()
            }
        }

        parent.addView(toggle.toggleView)
    }

    fun addTextbox(
        name: String?,
        description: String?,
        parent: LinearLayout,
        method: MethodPair<*, String>?,
        typeIs: String?
    ) {
        method ?: return
        val textbox: TextboxResult =  ComponentUtils.addTextbox(
            context,
            name,
            description,
            method,
            typeIs
        )

        parent.addView(textbox.textboxView)
    }

    fun addDropdown(
        id: String?,
        name: String?,
        description: String?,
        parent: LinearLayout,
        jsonArray: JSONArray?,
        method: MethodPair<*, String>?,
        controlledParentId: String?
    ) {
        method ?: return

        val dropdown =  ComponentUtils.addDropdown(
            context,
            name,
            description,
            jsonArray,
            method
        )

        if (id != null) {
            dropdownMap[id] = dropdown
        }

        if (controlledParentId != null) {
            childDropdownMap
                .getOrPut(controlledParentId) {
                    mutableListOf()
                }
                .add(dropdown)
        }

        if (controlledParentId != null) {
            val parentToggle = toggleMap[controlledParentId]

            if (parentToggle?.isChecked == true) {
                dropdown.enable?.run()
            } else {
                dropdown.disable?.run()
            }
        }

        parent.addView(dropdown.dropDownView)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val parentLayout: LinearLayout = view.findViewById(R.id.linear2)
        Integrations.addEveryPresets(requireContext(), parentLayout, this)
    }

    private fun aStyleButton1(button: LinearLayout?) {
        val drawable = GradientDrawable()
        drawable.cornerRadius = 30f
        drawable.setColor("#151515".toColorInt())
        button?.background = drawable
    }
}